### ISSUE TITLE
server.py | RuntimeError: OrderedDict mutated during iteration

### ISSUE THREAD

**Issue by janikvonrotz** [odoo/odoo#79823]
https://github.com/odoo/odoo/blob/58c1b04de1ec9179b940ee930984a232bb5de30d/odoo/service/server.py#L443

This line often throws a runtime error. I use docker and multiple dabases using the same cron thread. The registry dictionary.

```
RuntimeError: OrderedDict mutated during iteration
    for db_name, registry in registries.d.items():
  File "/usr/lib/python3/dist-packages/odoo/service/server.py", line 408, in cron_thread
    self.cron_thread(i)
  File "/usr/lib/python3/dist-packages/odoo/service/server.py", line 432, in target
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.7/threading.py", line 865, in run
```

Currently I deploy this patch:

```bash
#!/bin/bash

# Get script name
SCRIPT=$(basename "$0")

# Default configuration
PACKAGE_PATH="/usr/lib/python3/dist-packages/"

# Custom configuration
FILE_PATH="${PACKAGE_PATH}odoo/service/server.py"
FIND="for db_name, registry in registries.d.items()"
REPLACE="for db_name, registry in registries.d.copy().items()"

# Get filename
FILE=`basename "$FILE_PATH"`

# Cop from docker
docker cp "$1":"$FILE_PATH" .

# Find and replace
sed -i'' "s/$FIND/$REPLACE/g" $FILE

# Copy back to docker
docker cp $FILE "$1":"$FILE_PATH"

# Remove file
rm -f $FILE

# Register patch installation
echo "Patch $SCRIPT applied."
touch "/var/tmp/$SCRIPT"

```

Could we make this change permanent? I would welcome a discussion on this.

I think the cron job should never fail. The thread loop must be able to restart itself.

**Comment by dupski** [comment]
Just spent most of the day trying to track down why Odoo 14 would only run cron tasks once per server start on our production server, and arrived at the same issue! Glad we're not the only ones :)

The main problem is the `cron_thread()` function in `server.py` needs an exception handler just inside the `while True:` block, otherwise any errors cause the method to quit and no more crons to run until the server is restarted.

I ended up adding a lot more logging and exception handling to the `cron_thread()` function in `server.py` to make it more observable and resilient, as per below!:

```py
    def cron_thread(self, number):
        from odoo.addons.base.models.ir_cron import ir_cron
        while True:
            try:
                time.sleep(SLEEP_INTERVAL + number)     # Steve Reich timing style
                registries = odoo.modules.registry.Registry.registries
                _logger.info('cron%d polling for jobs', number)
                for db_name, registry in registries.d.items():
                    _logger.info('cron%d running for db: %s' % (number, db_name))
                    if registry.ready:
                        thread = threading.currentThread()
                        thread.start_time = time.time()
                        try:
                            ir_cron._acquire_job(db_name)
                        except Exception:
                            _logger.warning('cron%d encountered an Exception:', number, exc_info=True)
                        thread.start_time = None
                    else:
                        _logger.warning('cron%d db: %s: registry not ready.' % (number, db_name))
                _logger.info('cron%d finished cronloop.' % (number))
            except:
                _logger.error('cron%d exception.' % (number), exc_info=1)
```

Not sure why the list of databases is changing. We are running multi-tenant and do have a slightly complex db_filter set which could be contributing to the issue.

**Comment by janikvonrotz** [comment]
@dupski The `.items()` method returns a dict and at some point I assume the `registry` value changes.

I've setup a loki instance to monitor these issues, but did not have time to dive further into the topic.

It would be great to see a PR making this more resilient.

**Comment by janikvonrotz** [comment]
I have a new assumption that is error occurs when having multiple instance and only worker in a odoo docker container.

**Comment by sherpya** [comment]
I've got same problem with odoo 14, how can I track modifications o `registries.d`?

**Comment by janikvonrotz** [comment]
@sherpya I have not been able to reproduce the issue.

The best solution yes is to run `wokers=2` in the odoo config. This will require a proxy access.

**Comment by sherpya** [comment]
@janikvonrotz yes it's not easy to reproduce, I've got the problem on two instances and it started to happens recently.

An exeception in that place makes the cron thread to die without notice, I think it would a better
to make the function more robust and restart the loop.

The list() workaround does work anyway.

**Comment by jcardus** [comment]
Sorry for the mess guys, I'm trying to make a PR to fix this. I'm also having this problem in odoo 16 with multibd.

**Comment by Julien00859** [comment]
Running odoo in threaded mode isn't well suited for production. e.g. there is no process reaper that watch the many threads and kill them/restart them when needed (timeout/fatal errors/....). Such a process reaper exists in worker mode so I suggest you start odoo-bin with `--workers=x`.

Still, we agree this is an error and we'll look forward to solve it, I'm not too keen on the fix at #118862. My gut feeling is that there is a module install/upgrade/deletion at the same time that a cron is running (hence, inside of that loop). The registries could be updated and the loop would rightfully crash. If I'm right, then all the solutions to this problem are implemented in the worker mode already. Again, I suggest you don't use the threaded server in production.

**Comment by Julien00859** [comment]
I'm closing this issue then, if anyone still have the issue in worker mode, feel free to send a new message, we'll re-open it
